File: static/js/diagnostics.js
// ==============================================================================
//  AI Face Recognition Attendance System - System Diagnostics Test Runner
// ==============================================================================

let lastDiagnosticReport = null;

const initialTestDefinitions = [
    { id: 1, name: "Dependencies & Runtime Environment", desc: "Verifies Python 3.11, OpenCV, face_recognition, dlib, numpy, pandas, reportlab, psutil" },
    { id: 2, name: "Directory Structure & Write Permissions", desc: "Validates known_faces, attendance, unknown_faces, reports, and logs folders" },
    { id: 3, name: "Camera Stream Connectivity & FPS", desc: "Pings ESP32-CAM stream URL, checks buffer latency, and verifies video frame read" },
    { id: 4, name: "Face Engine & Encodings Cache", desc: "Validates encodings.pkl cache integrity, recognition tolerances, and model initialization" },
    { id: 5, name: "Attendance Engine & CSV Storage", desc: "Tests CSV record appending, schema integrity, and duplicate attendance debouncing" },
    { id: 6, name: "Registration Capture & Blur Pipeline", desc: "Validates Laplacian blur estimation, burst frame sampling, and multi-angle scoring" },
    { id: 7, name: "Anti-Spoofing & Liveness Verification", desc: "Tests Eye Aspect Ratio (EAR) blink detection, spectral texture, and multi-frame fusion" },
    { id: 8, name: "Unknown Face Registry & Conversion", desc: "Tests visitor auto-capturing, JSON metadata tracking, and student conversion" },
    { id: 9, name: "Multi-Format Exporters (CSV / XLSX / PDF)", desc: "Generates test daily/weekly CSV, styled Excel (.xlsx), and institutional PDF reports" },
    { id: 10, name: "Analytics Engine Aggregation", desc: "Validates KPI summaries, 7-day attendance trends, and peak arrival distribution" }
];

document.addEventListener("DOMContentLoaded", () => {
    renderInitialTestCards();
});

function renderInitialTestCards() {
    const container = document.getElementById("testItemsContainer");
    if (!container) return;

    container.innerHTML = initialTestDefinitions.map(t => `
        <div class="test-card" id="test-card-${t.id}">
            <div class="test-info">
                <div class="test-num">${String(t.id).padStart(2, '0')}</div>
                <div>
                    <div class="test-title">${t.name}</div>
                    <div class="test-desc" id="test-desc-${t.id}">${t.desc}</div>
                </div>
            </div>
            <div id="test-badge-${t.id}" class="test-badge badge-pending">
                <i class="fa-solid fa-clock"></i> PENDING
            </div>
        </div>
    `).join('');
}

async function runDiagnostics() {
    const runBtn = document.getElementById("runDiagBtn");
    const exportBtn = document.getElementById("exportDiagBtn");
    const statusBadge = document.getElementById("overallStatusBadge");
    const summaryText = document.getElementById("diagSummaryText");
    const lastRunTime = document.getElementById("lastRunTime");
    const consoleEl = document.getElementById("logConsole");

    if (runBtn) {
        runBtn.disabled = true;
        runBtn.innerHTML = '<i class="fa-solid fa-spinner fa-spin"></i> Running Diagnostics...';
    }
    if (exportBtn) exportBtn.disabled = true;

    // Reset badges to running
    for (let i = 1; i <= 10; i++) {
        const badge = document.getElementById(`test-badge-${i}`);
        if (badge) {
            badge.className = "test-badge badge-running";
            badge.innerHTML = '<i class="fa-solid fa-circle-notch fa-spin"></i> TESTING';
        }
    }

    if (statusBadge) {
        statusBadge.style.background = "rgba(99,102,241,0.2)";
        statusBadge.style.color = "#818cf8";
        statusBadge.innerHTML = '<i class="fa-solid fa-spinner fa-spin"></i> TESTING IN PROGRESS';
    }
    if (summaryText) summaryText.textContent = "Executing 10-Point System Diagnostics...";
    if (consoleEl) consoleEl.textContent = `[${new Date().toLocaleTimeString()}] Initiating full automated diagnostic test suite...\n`;

    try {
        const res = await fetch("/api/diagnostics/run");
        const data = await res.json();
        lastDiagnosticReport = data;

        if (data.status === "success" && data.report) {
            const report = data.report;
            
            // Update hero stats
            document.getElementById("pillPassed").textContent = report.passed;
            document.getElementById("pillFailed").textContent = report.failed;
            document.getElementById("pillDuration").textContent = `${report.total_duration_sec}s`;
            
            if (lastRunTime) lastRunTime.textContent = `Last run at ${report.timestamp} (Duration: ${report.total_duration_sec}s)`;

            let consoleOutput = `[${report.timestamp}] Full diagnostics completed in ${report.total_duration_sec}s\n`;
            consoleOutput += `Summary: ${report.passed}/${report.total_tests} Tests Passed\n`;
            consoleOutput += `------------------------------------------------------------\n`;

            // Update individual test cards
            report.tests.forEach(t => {
                const badge = document.getElementById(`test-badge-${t.id}`);
                const desc = document.getElementById(`test-desc-${t.id}`);
                
                if (badge) {
                    if (t.status === "PASS") {
                        badge.className = "test-badge badge-pass";
                        badge.innerHTML = `<i class="fa-solid fa-check"></i> PASS (${t.duration_ms}ms)`;
                    } else {
                        badge.className = "test-badge badge-fail";
                        badge.innerHTML = `<i class="fa-solid fa-xmark"></i> FAIL (${t.duration_ms}ms)`;
                    }
                }
                if (desc && t.details) {
                    desc.innerHTML = `<span style="color: ${t.status === 'PASS' ? '#a7f3d0' : '#fca5a5'}; font-family: 'JetBrains Mono', monospace; font-size: 0.78rem;">${t.details}</span>`;
                }

                consoleOutput += `[${String(t.id).padStart(2, '0')}] ${t.name.padEnd(42)} [${t.status}] (${t.duration_ms}ms)\n       -> ${t.details}\n`;
            });

            if (consoleEl) consoleEl.textContent = consoleOutput;

            // Overall badge
            if (statusBadge) {
                if (report.failed === 0) {
                    statusBadge.style.background = "rgba(16,185,129,0.2)";
                    statusBadge.style.color = "#10b981";
                    statusBadge.style.borderColor = "rgba(16,185,129,0.4)";
                    statusBadge.innerHTML = '<i class="fa-solid fa-circle-check"></i> 100% PASS - SYSTEM OPERATIONAL';
                    if (summaryText) summaryText.textContent = "All 10 Verification Checks Passed Successfully";
                } else {
                    statusBadge.style.background = "rgba(239,68,68,0.2)";
                    statusBadge.style.color = "#ef4444";
                    statusBadge.style.borderColor = "rgba(239,68,68,0.4)";
                    statusBadge.innerHTML = `<i class="fa-solid fa-circle-xmark"></i> ${report.failed} TESTS FAILED`;
                    if (summaryText) summaryText.textContent = `Diagnostics Detected ${report.failed} Issue(s)`;
                }
            }

            if (exportBtn) exportBtn.disabled = false;
        } else {
            alert("Diagnostics execution failed: " + (data.message || "Unknown error"));
        }
    } catch (err) {
        alert("Failed to connect to diagnostics API: " + err.message);
    } finally {
        if (runBtn) {
            runBtn.disabled = false;
            runBtn.innerHTML = '<i class="fa-solid fa-play"></i> Run Full System Test';
        }
    }
}

function exportDiagnosticReport() {
    if (!lastDiagnosticReport) return;
    const blob = new Blob([JSON.stringify(lastDiagnosticReport, null, 2)], { type: "application/json" });
    const url = URL.createObjectURL(blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = `Diagnostic_Report_${new Date().toISOString().slice(0,10)}.json`;
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
    URL.revokeObjectURL(url);
}

File: diagnostics.py
import os
import sys
import time
import json
import traceback
import numpy as np
import pandas as pd
from datetime import datetime

import config
from logger_util import log_system, log_error

class SystemDiagnostics:
    """
    Automated 10-Point End-to-End System Diagnostics & Health Verification Suite.
    """
    def __init__(self):
        self.results = []
        self.total_tests = 10
        self.passed_tests = 0
        self.failed_tests = 0

    def run_all(self) -> dict:
        """Executes all 10 diagnostic test suites and compiles detailed report."""
        self.results = []
        self.passed_tests = 0
        self.failed_tests = 0
        start_time = time.time()

        tests = [
            ("Dependencies & Runtime", self.test_dependencies),
            ("Directory Structure & Permissions", self.test_directories_and_permissions),
            ("Camera Stream Connectivity", self.test_camera_stream),
            ("Face Engine & Encodings Cache", self.test_face_engine),
            ("Attendance Engine & CSV Storage", self.test_attendance_engine),
            ("Registration Capture & Blur Pipeline", self.test_registration_pipeline),
            ("Anti-Spoofing & Liveness Verification", self.test_anti_spoofing),
            ("Unknown Face Registry & Conversion", self.test_unknown_face_manager),
            ("Multi-Format Exporters (CSV/XLSX/PDF)", self.test_report_exporters),
            ("Analytics Engine Aggregation", self.test_analytics_engine)
        ]

        for idx, (name, test_func) in enumerate(tests, 1):
            t_start = time.time()
            try:
                success, details = test_func()
                duration_ms = int((time.time() - t_start) * 1000)
                status = "PASS" if success else "FAIL"
                if success:
                    self.passed_tests += 1
                else:
                    self.failed_tests += 1
                    
                self.results.append({
                    "id": idx,
                    "name": name,
                    "status": status,
                    "duration_ms": duration_ms,
                    "details": details
                })
            except Exception as e:
                duration_ms = int((time.time() - t_start) * 1000)
                self.failed_tests += 1
                self.results.append({
                    "id": idx,
                    "name": name,
                    "status": "FAIL",
                    "duration_ms": duration_ms,
                    "details": f"Unhandled Exception: {str(e)}"
                })

        total_duration = round(time.time() - start_time, 2)
        overall_status = "ALL PASS" if self.failed_tests == 0 else f"{self.failed_tests} FAILED"

        report = {
            "timestamp": datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
            "overall_status": overall_status,
            "total_tests": self.total_tests,
            "passed": self.passed_tests,
            "failed": self.failed_tests,
            "total_duration_sec": total_duration,
            "tests": self.results
        }
        return report

    # ----------------- Individual Test Implementations -----------------

    def test_dependencies(self):
        """1. Verify all required enterprise Python packages are importable."""
        modules = [
            ("cv2", "OpenCV"),
            ("face_recognition", "face_recognition"),
            ("dlib", "dlib"),
            ("numpy", "NumPy"),
            ("pandas", "Pandas"),
            ("openpyxl", "openpyxl (Excel)"),
            ("reportlab", "ReportLab (PDF)"),
            ("psutil", "psutil (Telemetry)")
        ]
        missing = []
        for mod, name in modules:
            try:
                __import__(mod)
            except ImportError:
                missing.append(name)

        if missing:
            return False, f"Missing Python dependencies: {', '.join(missing)}"
        return True, f"All {len(modules)} core AI, computer vision, and export libraries verified."

    def test_directories_and_permissions(self):
        """2. Verify required storage directories exist and are writable."""
        config.ensure_directories()
        folders = [
            ("known_faces", config.KNOWN_FACES_DIR),
            ("attendance", config.ATTENDANCE_DIR),
            ("thumbnails", config.THUMBNAILS_DIR),
            ("unknown_faces", config.UNKNOWN_FACES_DIR),
            ("reports", config.REPORTS_DIR),
            ("logs", config.LOGS_DIR)
        ]
        for label, path in folders:
            if not os.path.exists(path):
                return False, f"Directory missing: {path}"
            # Test write permission
            test_file = os.path.join(path, ".perm_test.tmp")
            try:
                with open(test_file, "w") as f:
                    f.write("test")
                os.remove(test_file)
            except Exception as e:
                return False, f"Write permission denied for {label}: {str(e)}"

        return True, f"Verified write permissions across {len(folders)} enterprise storage paths."

    def test_camera_stream(self):
        """3. Test camera stream reachability and placeholder fallback."""
        import cv2
        from camera_stream import CameraStream
        
        stream = CameraStream(source=config.CAMERA_SOURCE)
        stream.start()
        time.sleep(0.4)
        
        frame = stream.get_frame()
        status_info = stream.get_status_info()
        stream.stop()

        if frame is None or frame.shape[0] == 0:
            return False, "Failed to retrieve video frame from CameraStream"

        status_text = status_info.get("status", "unknown")
        src = status_info.get("source", config.CAMERA_SOURCE)
        return True, f"Camera stream operational ({src}). Stream state: {status_text.upper()}."

    def test_face_engine(self):
        """4. Verify FaceEngine encodings cache and recognition tolerances."""
        from face_engine import FaceEngine
        from attendance_manager import AttendanceManager
        
        att_mgr = AttendanceManager()
        engine = FaceEngine(attendance_manager=att_mgr)
        
        cached_count = len(engine.known_face_encodings)
        unique_students = len(engine.known_face_names_unique)
        
        # Test synthetic frame processing
        dummy_frame = np.zeros((480, 640, 3), dtype=np.uint8)
        processed = engine.process_frame(dummy_frame)
        if processed is None:
            return False, "FaceEngine process_frame returned None"

        return True, f"Face engine verified (Tolerance={config.RECOGNITION_TOLERANCE}, FrameSkip={config.FRAME_SKIP}). Loaded {cached_count} encodings for {unique_students} students."

    def test_attendance_engine(self):
        """5. Verify attendance CSV generation, daily schema, and duplicate debounce."""
        from attendance_manager import AttendanceManager
        mgr = AttendanceManager()
        
        today = mgr._get_today_date_str()
        csv_path = mgr._get_csv_path(today)
        
        # Test mark
        test_student = "_DIAGNOSTIC_TEST_USER_"
        marked = mgr.mark_attendance(test_student)
        
        if not os.path.exists(csv_path):
            return False, f"Attendance CSV not created at {csv_path}"

        # Verify duplicate prevention
        dup_marked = mgr.mark_attendance(test_student)
        if dup_marked.get("status") == "success":
            return False, "Duplicate attendance was marked on the same day (duplicate prevention failed)"

        # Clean up test user from CSV
        try:
            df = pd.read_csv(csv_path)
            df = df[df['Name'] != test_student]
            df.to_csv(csv_path, index=False)
        except Exception:
            pass

        return True, f"Attendance engine verified. CSV created at Attendance_{today}.csv with duplicate prevention."

    def test_registration_pipeline(self):
        """6. Test blur score calculation and image burst quality evaluation."""
        from face_engine import FaceEngine
        engine = FaceEngine()

        # Sharp frame
        sharp_img = np.random.randint(0, 255, (100, 100, 3), dtype=np.uint8)
        sharp_blur = engine._calculate_blur_score(sharp_img)

        # Blurry flat frame
        flat_img = np.ones((100, 100, 3), dtype=np.uint8) * 128
        flat_blur = engine._calculate_blur_score(flat_img)

        if sharp_blur < flat_blur:
            return False, "Laplacian blur metric inverted"

        return True, f"Registration pipeline verified (Sharp={sharp_blur:.1f}, Blur={flat_blur:.1f}, MinThresh={config.BLUR_THRESHOLD})."

    def test_anti_spoofing(self):
        """7. Verify EAR blink detection, texture analysis, and liveness scoring."""
        from anti_spoofing import AntiSpoofingTracker
        tracker = AntiSpoofingTracker()

        # EAR test
        open_eye = [(10, 10), (13, 7), (17, 7), (20, 10), (17, 13), (13, 13)]
        closed_eye = [(10, 10), (13, 10), (17, 10), (20, 10), (17, 11), (13, 11)]
        ear_open = tracker.calculate_ear(open_eye)
        ear_closed = tracker.calculate_ear(closed_eye)

        if ear_open <= ear_closed or ear_closed > config.LIVENESS_EAR_THRESHOLD:
            return False, f"EAR calculation failed (Open: {ear_open:.2f}, Closed: {ear_closed:.2f})"

        # Texture test
        natural_img = np.random.randint(0, 255, (100, 100, 3), dtype=np.uint8)
        _, is_valid = tracker.analyze_texture(natural_img)
        if not is_valid:
            return False, "Natural texture rejected as spoof"

        return True, f"Anti-spoofing verified (Open EAR={ear_open:.2f}, Closed EAR={ear_closed:.2f}, Texture=PASS)."

    def test_unknown_face_manager(self):
        """8. Test unknown visitor logging and JSON registry."""
        from unknown_manager import UnknownFaceManager
        mgr = UnknownFaceManager()

        dummy_frame = np.random.randint(50, 200, (480, 640, 3), dtype=np.uint8)
        dummy_crop = np.random.randint(50, 200, (80, 80, 3), dtype=np.uint8)

        mgr.last_log_time = 0.0
        rec = mgr.record_unknown(dummy_frame, dummy_crop, confidence_score=35.0)
        if rec is None:
            return False, "Failed to record unknown visitor entry"

        rec_id = rec["id"]
        del_res = mgr.delete_record(rec_id)
        if not del_res.get("success"):
            return False, f"Failed to clean up test unknown record {rec_id}"

        return True, "Unknown face auto-capture, JSON registry, and record deletion verified."

    def test_report_exporters(self):
        """9. Test CSV, Excel (.xlsx), and PDF (.pdf) generation."""
        from reports_generator import ReportsGenerator
        from attendance_manager import AttendanceManager
        
        rep_gen = ReportsGenerator(attendance_manager=AttendanceManager())
        
        csv_p, _ = rep_gen.generate_csv(report_type="daily")
        xlsx_p, _ = rep_gen.generate_excel(report_type="daily")
        pdf_p, _ = rep_gen.generate_pdf(report_type="daily")

        for p, fmt in [(csv_p, "CSV"), (xlsx_p, "Excel"), (pdf_p, "PDF")]:
            if not os.path.exists(p) or os.path.getsize(p) == 0:
                return False, f"{fmt} report generation failed or output file is empty ({p})"

        return True, "Verified CSV, styled Excel (.xlsx), and institutional PDF (.pdf) report generators."

    def test_analytics_engine(self):
        """10. Test analytics KPI calculation, trends, and arrival distribution."""
        from analytics_engine import AnalyticsEngine
        from attendance_manager import AttendanceManager
        from unknown_manager import UnknownFaceManager
        from anti_spoofing import AntiSpoofingTracker

        analytics = AnalyticsEngine(
            attendance_manager=AttendanceManager(),
            unknown_manager=UnknownFaceManager(),
            anti_spoofing=AntiSpoofingTracker()
        )

        summary = analytics.get_summary_metrics()
        trends = analytics.get_trends_data(7)
        hourly = analytics.get_hourly_distribution()

        if "total_students" not in summary or "labels" not in trends or "labels" not in hourly:
            return False, "Analytics engine returned incomplete data structure"

        return True, f"Analytics engine verified. Processed KPIs, 7-day trend series, and hourly distribution."

if __name__ == "__main__":
    print("\n=======================================================")
    print("  AI Face Recognition Attendance - Full System Diagnostics")
    print("=======================================================\n")
    diag = SystemDiagnostics()
    report = diag.run_all()

    for t in report["tests"]:
        status_color = "\033[92mPASS\033[0m" if t["status"] == "PASS" else "\033[91mFAIL\033[0m"
        print(f"[{t['id']:02d}/10] {t['name']:<40} [{status_color}] ({t['duration_ms']} ms)")
        print(f"       -> {t['details']}")

    print("\n-------------------------------------------------------")
    print(f"  Summary: {report['passed']}/{report['total_tests']} PASSED | Total Time: {report['total_duration_sec']}s")
    print("=======================================================\n")
